File: app/src/main/java/com/hana053/micropost/ui/pages/signup/email/SignupEmailFragment.java
package com.hana053.micropost.ui.pages.signup.email;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.hana053.micropost.R;
import com.hana053.micropost.databinding.SignupEmailBinding;
import com.hana053.micropost.ui.pages.signup.SignupBaseFragment;
import com.hana053.micropost.ui.pages.signup.SignupViewModel;

import lombok.NoArgsConstructor;

@NoArgsConstructor
public class SignupEmailFragment extends SignupBaseFragment<SignupEmailBinding> implements SignupEmailViewListener {

    public static Fragment newInstance(SignupViewModel viewModel) {
        final SignupEmailFragment fragment = new SignupEmailFragment();
        fragment.setViewModelAsArguments(viewModel);
        return fragment;
    }

    @Override
    public View.OnClickListener onClickNextBtn() {
        return nextBtnHandler.call(signupCtrl -> {
            signupCtrl.navigateToNewPassword();
            return true;
        });
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.signup_email, container, false);
    }

    @Override
    protected SignupEmailBinding setupBinding(SignupViewModel viewModel) {
        SignupEmailBinding binding = SignupEmailBinding.bind(getView());
        binding.setViewModel(viewModel);
        binding.setListener(this);
        binding.setBtnHandler(getBtnHandler());
        return binding;
    }

    @Override
    protected boolean isFormValid() {
        final String emailPattern = "^[^0-9][a-zA-Z0-9_]+([.][a-zA-Z0-9_]+)*@\\[?([\\d\\w\\.-]+)]?$";
        return getEmail().matches(emailPattern);
    }

    @Override
    protected void showOrHideErrorMsg(boolean isValid) {
        if (isValid || getEmail().length() == 0)
            getBinding().emailInvalid.setVisibility(View.GONE);
        else
            getBinding().emailInvalid.setVisibility(View.VISIBLE);
    }

    @NonNull
    private String getEmail() {
        return getViewModel().email.get();
    }

}

File: app/src/main/kotlin/com/hana053/micropost/pages/signup/email/SignupEmailFragment.kt
package com.hana053.micropost.pages.signup.email

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.github.salomonbrys.kodein.KodeinInjector
import com.github.salomonbrys.kodein.android.SupportFragmentInjector
import com.github.salomonbrys.kodein.instance
import com.hana053.micropost.R
import com.trello.rxlifecycle.components.support.RxFragment


class SignupEmailFragment : RxFragment(), SupportFragmentInjector {

    override val injector: KodeinInjector = KodeinInjector()

    private val presenter: SignupEmailPresenter by instance()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_signup_email, container, false)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        initializeInjector()

        presenter.bind()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        destroyInjector()
    }

}